### DISCUSSION TITLE
Vercel Edge Middleware Request Body?

### DISCUSSION THREAD

**Discussion by cj** [vercel/community#78] [Help] [answered]
How do you access the request body is Edge Middleware?  `req.json()`a and `req.text()` are always empty.

**Comment by cj** [comment]
It looks like it was also brought up here https://github.com/vercel/next.js/discussions/30668

**Reply by Pieparker** [reply]
Thank you for cross-posting the discussion in the Next.js repo. Given that this is relevant to Middleware and Next.js (and not a problem for edge functions as a whole) and in order to keep the conversation focused in a single setting, we'll follow along with any updates in https://github.com/vercel/next.js/discussions/30668.

**Reply by cj** [reply]
Hi @Pieparker thank you for the reply, and I hope you're having a great weekend!

I am not sure I 100% agree it's just a Nextj.js issue and not the Edge functions as a whole.  I have looked into the Nextjs code https://github.com/vercel/next.js/blob/canary/packages/next/server/web/adapter.ts#L10-L35 and https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/loaders/next-middleware-loader.ts, it is consuming the Vercel Edge Functions (Cloudflare Workers) directly.   As bare metal Cloudflare Workers provide direct access to the request body https://developers.cloudflare.com/workers/examples/read-post, I believe that the Vercel Edge Functions are not forwarding the body along for whatever reason.  Another thing I noticed is there are no examples for consuming the body of a request https://github.com/vercel/examples/tree/main/edge-functions, so is the body not being forwarded from CF Workers intended?  If it is, I would love to know the reasoning behind the decision as I feel like it limits what can be accomplished with Vercel Edge Functions.

Cheers!

**Reply by Pieparker** [reply]
I've double-checked and it is intentional: middleware doesn't accept data (i.e. in a POST request) due to complexities with multiple middleware handlers. However, you could fall through from middleware to a serverless function via a rewrite!

**Reply by cj** [reply]
@Pieparker thank you for confirming.

I would love to suggest the original body get passed through as read-only. That way, the middleware could be used for simple API requests without falling back to serverless functions (which are slower), and it would avoid any issues with multiple middleware handlers.

Cheers!

**Reply by neoromantic** [reply]
It was a long-awaited improvement: to be able to process some requests without very slow cold-boot related overhead. 

I.e. I need to make a simple response to some POST request to my function. I can't do it on edge functions. Only offer Vercel gives me is an API function, which could respond up to 2(!) seconds, which is extremely slow and unsuitable for many needs.

I second the suggestion to allow read-only access for whole request object.

**Reply by Pieparker** [reply]
Appreciate this, both of you. What I'll do is convert this to a feature suggestion and we can track accordingly!

**Reply by malaba6** [reply]
@Pieparker Any update on this? coz I'm running into the same issue. Thanks!

**Reply by 0x2me** [reply]
Also running into this issue, any update ?

**Reply by Chukwu3meka** [reply]
hi @Pieparker, Please do you have any update

**Reply by morshedgit** [reply]
I do like this feature as well!

**Reply by filipedeschamps** [reply]
Would be lovely to have access to the `body` to make some fine-grained rate limit.

**Reply by SawkaDev** [reply]
@Pieparker  hi, any update on this or do you have any resources for a workaround? This is useful to create Audit Logs of all CRUD operations

**Reply by amyegan** [reply]
Hi Everyone! In [Next.js v12.1.1](https://github.com/vercel/next.js/releases/tag/v12.1.1) support was added for reading request bodies in middleware (https://github.com/vercel/next.js/pull/34519). Can you confirm this still isn't working for you?

**Reply by SawkaDev** [reply]
@amyegan Yes I am able to view the request body now! However request body is of type `ReadableStream`. So once I convert the ReadableStream to a string i can view the body properly! Thanks you.

However, I am still not able to view the response body. Is that expected?

**Reply by amyegan** [reply]
Hi @Sliffcak. If you are trying to add data to the body of a response from Middleware, that is not supported in Next.js. In that case, falling through to a serverless function is the recommended solution. You can learn more here: [Returning response body in Middleware](https://nextjs.org/docs/messages/returning-response-body-in-middleware). 

Another hack for passing data from Middleware to the page is adding the data to a query parameter (searchParams) in the URL which you can read from `getServerSideProps`.

If I misunderstood what you were asking about the response body, please let me know. I want to make sure you get the right information!

**Reply by SawkaDev** [reply]
@amyegan  thanks for the follow up. I am not trying to have the middleware create a response body but rather I want to view the contents of the NextResponse

```
export async function middleware(request: NextRequest) {
  // for now lets assume we know the API being contacted will return a json body
  const response = NextResponse.next(); 
  console.log(request.body);  // valid data
  console.log(response.body); // undefined

  return response;
}
```

The use case for this is for logging reasons. I want to use middleware to log the request & response of specific API endpoints. 

I am guessing I am trying something that middlware is not meant for. I was just hoping to gain access to the response that the request made inside of the middleware.

**Comment by kay-is** [comment, marked as answer]
I used this hack: 

    const json = await new Response(request.body).json();

**Reply by leebenson** [reply]
Thanks @kay-is ! This appears to be the only referenced fix _anywhere_. Thank you!

**Reply by SawkaDev** [reply]
can you post a full example of a middleware using this so everyone can have fully clarification? much appreciated.

**Reply by AidanGuarniere** [reply]
you could also use const json = (await req.json())

**Comment by la3rence** [comment]
In my case, my edge fuction still needs to use `request.body` after it being read. And the error logs shows `Response body object should not be disturbed or locked`.

Here's my solution:

```js
 const [bodyStream, bodyStreamClone] = req.body.tee();
```

It returns two instances cloned from the original request body. I use the second one to get and parse JSON data. See [MDN - tee()](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/tee).
